echeancier sur prestashop au choix de paiement

noshette opened the pull request:
Testé sur : 

v.1.5.6.2
Pas de message dans le paiement c'est lié à la version ?
L’icône fontawesome n'apparaît pas à coté de la méthode de paiement Alma

v.1.6.1.23
Pas de message dans le paiement c'est lié à la version ?

v.1.7.6.8
Ok

olance commented:
@noshette tu veux dire quoi par "Pas de message dans le paiement c'est lié à la version ?" ? ^^

noshette commented:
> @noshette tu veux dire quoi par "Pas de message dans le paiement c'est lié à la version ?" ? ^^

Pas très claire effectivement ^^

Quand on choisit la méthode de paiement, sur les versions < 1.7, et quand on y sélectionne Alma, ça redirige direct vers notre page de paiement, y'a pas de radio button avec la description qui apparaît (la ou j'ai placé l'échéancier)

olance commented:
Ah yes indeed, PrestaShop avant 1.7 effectivement c'était fait comme ça !
Le template pour le rendu du bouton de la méthode de paiement est différent, faut qu'on regarde ensemble comment on peut intégrer l'échéancier là-dedans :)

noshette commented:
Finalement j'ai trouvé comment afficher l'échéancier sur ps < 1.7 ;)

Il reste juste cette histoire d’icône moche et on est bon 🥇

olance commented on alma/controllers/hook/displayPayment.php:
```diff
@@ -80,9 +80,15 @@ public function run($params)
                 }
             }
 
+            $alma = AlmaClient::defaultInstance();
+            $paymentDataPnX = PaymentData::dataFromCart($this->context->cart, $this->context, $n);
+            $eligibilityPnX = $alma->payments->eligibility($paymentDataPnX);  
```
I think we need some heavy refactoring here:

- `AlmaEligibilityHelper::eligibilityCheck` should be the one to loop on `$n` and check eligibility for each activated installments count:
    - Get plans activated by the merchant
    - Check amount bounds for each activated plans: skip those for which amount is outside of bounds
    - Call Alma eligibility API for activated plans that should be active for cart amount
    - Return eligibility results

Something like this:
```
$eligibilities = [];
$activePlans = [];
foreach (AlmaSettings::activatedInstallmentsPlans() as $n) { // [2,3,4]
	if ($purchaseAmount < AlmaSettings::installmentPlanMinAmount($n) || $purchaseAmount > AlmaSettings::installmentPlanMaxAmount($n)) {
		$eligibility = new stdClass();
		$eligibility->installments_count = $n;
		$eligibility->isEligible = false;
		$eligibility->constraints = array(
			"purchase_amount" => array(
				"minimum" => AlmaSettings::installmentPlanMinAmount($n),
				"maximum" => AlmaSettings::installmentPlanMaxAmount($n)
			)
		);

		$eligibilities[] = $eligibility;
	} else {
		$activePlans[] = $n;
	}
}

$paymentData = PaymentData::dataFromCart($context->cart, $context, $activePlans);
$almaEligibilities = $alma->payments->eligibility($paymentData);

$eligibilities = array_merge($eligibilities, $almaEligibilities);

function cmp_installments_count($a, $b) {
	return $a->installments_count > $b->installments_count ? 1 : ($a->installments_count == $b->installments_count ? 0 : -1);
}

usort($eligibilities, 'cmp_installments_count');
return $eligibilities;
```

- Loop on each eligibility result and and display only a button for the eligible ones

olance replied:
Do check Shoping cart footer hook controller & paymentOptions hook controller to adapt them to the new eligibility check

olance commented on alma/controllers/hook/paymentOptions.php:
```diff
@@ -95,9 +96,13 @@ public function run($params)
             );
 
+            
+            $paymentDataPnX = PaymentData::dataFromCart($this->context->cart, $this->context, $n);
+            $eligibilityPnX = $alma->payments->eligibility($paymentDataPnX);            
             if (!$forEUComplianceModule && !empty(AlmaSettings::getPaymentButtonDescription())) {
                 $this->context->smarty->assign(array(
                     'desc' => sprintf(AlmaSettings::getPaymentButtonDescription(), $n),
+                    'plans' => (array) $eligibilityPnX->paymentPlan,
```
See previous comment :)

olance commented on alma/views/css/alma.css:
```diff
@@ -81,6 +81,21 @@ p.payment_module a.alma-button:after {
+    padding: 4px 0px;
+
+}
+.alma-p-fee-plan--descritpion{
```
```suggestion
.alma-p-fee-plan--description{
```

olance commented on alma/views/templates/hook/payment_button_desc.tpl:
```diff
@@ -24,3 +24,12 @@
 </p>
+{foreach from=$plans item=v name=counter}        
+        
+        <p class="alma-p-fee-plan--descritpion">    
```
```suggestion
        <p class="alma-p-fee-plan--description">    
```

olance commented on alma/views/templates/hook/payment_button_desc.tpl:
```diff
@@ -24,3 +24,12 @@
+{foreach from=$plans item=v name=counter}        
+        
+        <p class="alma-p-fee-plan--descritpion">    
+            <span>Echéance {$smarty.foreach.counter.iteration} : le {$v.due_date|date_format:"%d/%m/%Y"}</span>
+            <span>{math equation=$v.purchase_amount / 100 format="%.2f"}&euro;</span>
+        </p>
+         
+    {/foreach}
+    <br />
```
identation issues?

olance replied:
+ same display changes as in displayPayment.tpl

👉 could we refactor the installments display in a single tpl file, included by both "master" templates?

olance reviewed (changes requested):
See comments :)

olance commented on alma/controllers/hook/displayPayment.php:
```diff
@@ -36,77 +36,49 @@ class AlmaDisplayPaymentController extends AlmaProtectedHookController
+                if (AlmaSettings::showDisabledButton()) {
+                    $disabled = true;
+                    $plans = null;
+                }             
```
Add:

```php
else {
    continue;
}
```

olance commented on alma/controllers/hook/displayPayment.php:
```diff
@@ -36,77 +36,49 @@ class AlmaDisplayPaymentController extends AlmaProtectedHookController
 
-        $cart = $params['cart'];
-        $purchaseAmount = almaPriceToCents((float) $cart->getordertotal(true, Cart::BOTH));
+        $installmentPlans = AlmaEligibilityHelper::eligibilityCheck($this->context);
```
Do something when `$installmentPlans` is empty:

- Easy option: return
- Or: if disabled button display is enabled, loop over merchant-activated installments plan and show a disabled button for each (a payment option each with `$plans => null, $disabled => true, $error > true`)

olance commented on alma/views/templates/hook/displayPayment.tpl:
```diff
@@ -21,33 +21,43 @@
-                            {/if}
+{assign var="icone_diplsay" value=""}
+{if $prestashop_version < 1.6}
+    {assign var="icone_diplsay" value="disable_arrow_icone"}
```
```suggestion
    {assign var="iconDisplay" value="disable-arrow-icon"}
```

olance commented on alma/controllers/hook/displayPayment.php:
```diff
@@ -36,77 +36,49 @@ class AlmaDisplayPaymentController extends AlmaProtectedHookController
                 'desc' => sprintf(AlmaSettings::getPaymentButtonDescription(), 3),
                 'order_total' => (float) $cart->getOrderTotal(true, Cart::BOTH),
                 'options' => $options,
+                'prestashop_version' => _PS_VERSION_,
```
Should we rather use a boolean like `displayArrayIcon => version_compare(...)`

olance commented on alma/controllers/hook/displayShoppingCartFooter.php:
```diff
@@ -39,30 +39,44 @@ public function run($params)
+        $eligible = false;
+        foreach($eligibilities as $eligibility){
+            if($eligibility->isEligible){
+                $eligible = true;
```
```suggestion
                $eligible = true;
                break;
```

olance commented on alma/controllers/hook/displayShoppingCartFooter.php:
```diff
@@ -39,30 +39,44 @@ public function run($params)
-                        Tools::displayPrice(almaPriceFromCents($minAmount))
-                    );
-                }
+            $minimum = 9999999;
```
```suggestion
            $minimum = PHP_INT_MAX;
```

olance commented on alma/controllers/hook/displayShoppingCartFooter.php:
```diff
@@ -39,30 +39,44 @@ public function run($params)
+            $minimum = 9999999;
+            $maximum = 0;
+            foreach($eligibilities as $eligibility){
+                if(!$eligibility->isEligible){
```
Since we're in `if(!$eligible)`, this check is not necessary

olance commented on alma/includes/AlmaEligibilityHelper.php:
```diff
@@ -66,48 +66,74 @@ private static function checkPnXBounds($cart)
+            return array();
+        }
+        $n = 1;
+        while ($n < AlmaSettings::installmentPlansMaxN()) {
+            ++$n;
+            if (!AlmaSettings::isInstallmentPlanEnabled($n)) {
+                continue;
+            } else {
```
Add this to `AlmaSettings`:

```php
    public static function activeInstallmentsCounts()
    {
        $installmentsCounts = array();

        for ($n = 2; $n <= self::installmentPlansMaxN(); $n++) {
            if (self::isInstallmentPlanEnabled($n)) {
                $installmentsCounts[] = $n;
            }
        }

        return $installmentsCounts;
    }
```

and use `foreach(AlmaSettings::activeInstallmentsCounts() as $n)` :)

olance commented on alma/includes/AlmaEligibilityHelper.php:
```diff
@@ -66,48 +66,74 @@ private static function checkPnXBounds($cart)
             return null;
         }
+        
+        # hack installments plan 3 #
+        # API return p3x by default so it will be removed from almaEligibilities if it's in eligibilities
+        foreach($eligibilities as $eli){
+            if(3 === $eli->installmentsCount){
+                foreach($almaEligibilities as $k => $eligibility){
+                    if(3 === $eligibility->installmentsCount){
+                        unset($almaEligibilities[$k]);
+                    }
+                }
+            }
+        }
```
Instead, just don't call `$alma->payments->eligibility` if `$activePlans` is empty :)

olance commented on alma/controllers/hook/displayShoppingCartFooter.php:
```diff
@@ -39,30 +39,44 @@ public function run($params)
+                    $minAmount = $eligibility->constraints['purchase_amount']['minimum'];
+                    $maxAmount = $eligibility->constraints['purchase_amount']['maximum'];
+                    if ($cartTotal < $minAmount || $cartTotal > $maxAmount) {
+                        if ($cartTotal > $maxAmount && $maxAmount > $maximum) {
+                            $eligibilityMsg = ' ' . sprintf(
+                                $this->module->l('(Maximum amount: %s)', 'displayShoppingCartFooter'),
+                                Tools::displayPrice(almaPriceFromCents($maxAmount))
+                            );
+                            $maximum = $maxAmount;
+                        }
+                        if ($cartTotal < $minAmount && $minAmount < $minimum) {
+                            $eligibilityMsg = ' ' . sprintf(
```
Use the loop only to identify the global min/max values, then compute the message string after the loop depending on whether `$cartTotal` is higher than global max or lower than global min

olance commented on alma/includes/AlmaEligibilityHelper.php:
```diff
@@ -66,48 +66,74 @@ private static function checkPnXBounds($cart)
-
+        $eligibilities = array();
+        $activePlans = array();
+        $purchaseAmount = (float) Tools::ps_round((float)$context->cart->getOrderTotal(true, Cart::BOTH), 2) * 100;
```
Use `almaPriceToCents`

olance commented on alma/includes/AlmaEligibilityHelper.php:
```diff
@@ -66,48 +66,74 @@ private static function checkPnXBounds($cart)
+        $paymentData = PaymentData::dataFromCart($context->cart, $context, $activePlans);
+        if (!$paymentData) {
+            AlmaLogger::instance()->error('Cannot check cart eligibility: no data extracted from cart');
             return null;
```
```suggestion
            return array();
```

olance commented on alma/includes/AlmaEligibilityHelper.php:
```diff
@@ -66,48 +66,74 @@ private static function checkPnXBounds($cart)
                 "Error when checking cart {$context->cart->id} eligibility: " . $e->getMessage()
             );
-
             return null;
```
```suggestion
            return array();
```

olance commented on alma/includes/AlmaEligibilityHelper.php:
```diff
@@ -66,48 +66,74 @@ private static function checkPnXBounds($cart)
+                continue;
+            } else {
+                if ($purchaseAmount < AlmaSettings::installmentPlanMinAmount($n) || $purchaseAmount > AlmaSettings::installmentPlanMaxAmount($n)) {                    
+                    $eligibility = new stdClass();
```
Explore the option of creating a new `Alma\API\Endpoints\Results\Eligibility` ?

olance commented on alma/includes/PaymentData.php:
```diff
@@ -106,7 +108,7 @@ public static function dataFromCart($cart, $context, $installmentsCount = 3)
         $data = array(
             'payment' => array(
-                'installments_count' => (int) $installmentsCount,
+                'installments_count' => $installmentsCount,
```
Document $installmentsCount type in method's PHPDoc

olance commented on alma/views/css/alma.css:
```diff
@@ -58,6 +58,11 @@ p.payment_module a.alma-button:after {
     color: #777;
 }
 
+p.payment_module a.alma-button.disable_arrow_icone:after {
```
```suggestion
p.payment_module a.alma-button.disable-arrow-icon:after {
```

olance commented on alma/views/templates/hook/_partials/feePlan.tpl:
```diff
@@ -0,0 +1,36 @@
+            {if $smarty.foreach.counter.iteration === 1}
+                {l s='Today' mod='alma'}
+            {else}
+                {dateFormat date= $v.due_date|date_format:"%Y-%m-%d" full = 0}
```
```suggestion
                {dateFormat date=$v.due_date|date_format:"%Y-%m-%d" full=0}
```

olance commented on alma/views/templates/hook/displayPayment.tpl:
```diff
@@ -21,33 +21,43 @@
-                            {else}
-                                {l s='Alma Monthly Installments are not available for this order' mod='alma'}
-                            {/if}
+{assign var="icone_diplsay" value=""}
```
```suggestion
{assign var="iconDisplay" value=""}
```

olance commented on alma/views/templates/hook/displayPayment.tpl:
```diff
@@ -21,33 +21,43 @@
+        <div class="row">
+            <div class="col-xs-12">
+                <p class="payment_module">
+                    <a href="#" onclick="return false;" class="disabled alma-button  {$icone_diplsay}">
```
```suggestion
                    <a href="#" onclick="return false;" class="disabled alma-button {$iconDisplay}">
```

olance commented on alma/views/templates/hook/displayPayment.tpl:
```diff
@@ -21,33 +21,43 @@
             <div class="col-xs-12">
                 <p class="payment_module">
-                    <a href="{$option.link}" class="alma-button">
+                    <a href="{$option.link}" class="alma-button  {$icone_diplsay}">
```
```suggestion
                    <a href="{$option.link}" class="alma-button {$iconDisplay}">
```

noshette commented:
En pj les screenshots :  
- des 3 échéanciers 
- des 'notices' php à l'install du module 
<img width="741" alt="ps1-7" src="https://user-images.githubusercontent.com/3249959/100639238-34613080-3335-11eb-96e2-99ca42ca3846.png">
<img width="623" alt="notice-php" src="https://user-images.githubusercontent.com/3249959/100639246-362af400-3335-11eb-9b0f-a00bbb80718d.png">
<img width="752" alt="ps1-5" src="https://user-images.githubusercontent.com/3249959/100639248-37f4b780-3335-11eb-845d-a39cf3169ebc.png">
<img width="1192" alt="ps1-6" src="https://user-images.githubusercontent.com/3249959/100639249-388d4e00-3335-11eb-8ef9-718e681d287b.png">

olance commented on alma/controllers/hook/displayPayment.php:
```diff
@@ -75,23 +69,30 @@ public function run($params)
             }
+
+            if (is_callable('Media::getMediaPath')) {
+                $logo = Media::getMediaPath(_PS_MODULE_DIR_ . $this->module->name . '/views/img/logos/alma_p'.$n.'x.svg');
```
```suggestion
                $logo = Media::getMediaPath(_PS_MODULE_DIR_ . $this->module->name . "/views/img/logos/alma_p${n}x.svg");
```

olance commented on alma/controllers/hook/displayPayment.php:
```diff
@@ -75,23 +69,30 @@ public function run($params)
+            if (is_callable('Media::getMediaPath')) {
+                $logo = Media::getMediaPath(_PS_MODULE_DIR_ . $this->module->name . '/views/img/logos/alma_p'.$n.'x.svg');
+            } else {
+                $logo = $this->module->getPathUri() . '/views/img/logos/alma_p'.$n.'x.svg';
```
```suggestion
                $logo = $this->module->getPathUri() . "/views/img/logos/alma_p${n}x.svg";
```

olance commented on alma/controllers/hook/paymentOptions.php:
```diff
@@ -83,7 +84,7 @@ private function createPaymentOption($forEUComplianceModule, $ctaText, $action)
         } else {
             $paymentOption = new PrestaShop\PrestaShop\Core\Payment\PaymentOption();
-            $logo = Media::getMediaPath($baseDir . '/views/img/tiny_alma_payment_logos.svg');
+            $logo = Media::getMediaPath($baseDir . '/views/img/logos/alma_p'.$n.'x.svg');
```
```suggestion
            $logo = Media::getMediaPath("${baseDir}/views/img/logos/alma_p${n}x.svg");
```

noshette commented:
I replaced Alma logo everywhere and changed css for feesplans in payment page : 

# in module list
<img width="251" alt="modulelist1_7" src="https://user-images.githubusercontent.com/3249959/101053011-9022f800-3587-11eb-85b7-576d6a4ced5a.png">

<img width="890" alt="moduleliste1_6" src="https://user-images.githubusercontent.com/3249959/101052998-8e593480-3587-11eb-8026-19c22a00a66a.png">

# in Alma module
<img width="1183" alt="almamodule" src="https://user-images.githubusercontent.com/3249959/101053007-8f8a6180-3587-11eb-90c2-65319f6ebf30.png">

# in footer
<img width="801" alt="footer" src="https://user-images.githubusercontent.com/3249959/101053005-8f8a6180-3587-11eb-8cdd-517247fa7c50.png">

# in payment page
<img width="778" alt="payment1_5" src="https://user-images.githubusercontent.com/3249959/101052987-8b5e4400-3587-11eb-94ad-479a7b9f8ac3.png">

<img width="1212" alt="payment1_6" src="https://user-images.githubusercontent.com/3249959/101052991-8c8f7100-3587-11eb-85bf-48d4157277ed.png">

<img width="736" alt="payment1_7" src="https://user-images.githubusercontent.com/3249959/101053002-8ef1cb00-3587-11eb-8388-ff6030c6608b.png">
